crates/metadata-store/src/lib.rs

```diff
@@ -70,26 +58,27 @@ pub enum Precondition {
 pub trait MetadataStore {
     /// Gets the value and its current version for the given key. If key-value pair is not present,
     /// then return [`None`].
-    async fn get(&self, key: &str) -> Result<Option<VersionedValue>, ReadError>;
+    async fn get(&self, key: ByteString) -> Result<Option<VersionedValue>, ReadError>;
 
     /// Gets the current version for the given key. If key-value pair is not present, then return
     /// [`None`].
-    async fn get_version(&self, key: &str) -> Result<Option<Version>, ReadError>;
+    async fn get_version(&self, key: ByteString) -> Result<Option<Version>, ReadError>;
 
     /// Puts the versioned value under the given key following the provided precondition. If the
     /// precondition is not met, then the operation returns a [`WriteError::PreconditionViolation`].
     async fn put(
         &self,
-        key: &str,
+        key: ByteString,
```

Comment:
The reason why I switched to `ByteString` is that I needed an owned type for the `restate_node_protocol::metadata_store::Put` message and I wanted to avoid an extra clone operation for `&'static str` keys. With `ByteString::from_static(&'static str)` this should be possible. Moreover, I didn't want to change the traits type to `&'static str` which seemed too limiting for users. This is probably premature optimization at its best 😅

Reply:
Thanks for explaining. I think at some point we should explore using `smol_str` or `smartstring` for most of our internal strings.